### PR TITLE
Simple fix for vscode-reasonml#45

### PR DESCRIPTION
**Pull request by wilonth**
Fix reasonml-editor/vscode-reasonml#45:
- Let Merlin diagnostics go first, bsb's diagnostics may then override it when the file is saved
- Catch External Preprocessor error (i.e from Refmt) in bsbOutput error parser

### CODE REVIEW

**Comment by freebroccolo** [comment]
Thanks for this!

So this will (via `bsb`) actually remove diagnostics reported by merlin which have gone stale? How does that work? I guess this is simply overwriting the entire project diagnostics with the `bsb` output when a file is saved?

If so, I suppose this could work but I wonder if this would make the diagnostics experience feel a bit jarring or something like that. I'll have to give it a try.

@chenglou @jchavarri thoughts on this?

File: src/shared/parser/bucklescript/index.ts [hunk 1]
```diff
@@ -94,5 +94,33 @@ export function parseErrors(bsbOutput: string): { [key: string]: types.Diagnosti
...
+      message,
+      startCharacter,
+      startLine,
+      endCharacter,
+      endLine,
+      severity
+    );
+    if (!parsedDiagnostics[fileUri]) {
+      parsedDiagnostics[fileUri] = [];
+    }
```
Comment (freebroccolo) [inline src/shared/parser/bucklescript/index.ts]: Could we change this to the following?

```
parsedDiagnostics[fileUri] = parsedDiagnostics[fileUri] || [];
```

Comment (chenglou) [reply src/shared/parser/bucklescript/index.ts]: Careful with range handling in ocaml. The col can be either -1, 0, 1 or something weird. Here's the right way: https://github.com/BuckleScript/bucklescript/blob/b72cad66d205b608bddb94038fe2267da6f7a009/jscomp/super_errors/super_location.ml#L81

**Comment by chenglou** [comment]
I'm gonna cc @jchavarri on this one since he did the bsb integration and might have an opinion.

My views are here: https://github.com/freebroccolo/ocaml-language-server/pull/32#issuecomment-330023003. To summarize, bsb's report is 100% correct, so when it says there's no error (warnings is another matter) in a file, there really isn't. When it says there is, there definitely is (merlin on the other hand might give false positives/negatives as a tradeoff for speed and diagnosis after error recovery).

I think for at least files that are not the currently active one, we should override their merlin diagnosis (again, only for errors) with whatever the bsb one shows/doesn't show.

**Comment by freebroccolo** [comment]
> To summarize, bsb's report is 100% correct, so when it says there's no error (warnings is another matter) in a file, there really isn't.

Yeah, I'm not entirely sure how this is supposed to work without trying it, but what I am worried about is hiding some stuff reported by Merlin that `bsb` maybe doesn't even deal with, like certain warnings. I'm perfectly fine with replacing the stuff that overlaps though like the errors. It seems like we will probably need to do some sort of merging and pruning to get the ideal solution…

File: src/server/session/analyzer.ts [hunk 2]
```diff
@@ -46,6 +46,17 @@ export default class Analyzer implements rpc.Disposable {
...
   public refreshWithKind(syncKind: server.TextDocumentSyncKind): (id: types.TextDocumentIdentifier) => Promise<void> {
     return async (id) => {
+      
+      if (syncKind === server.TextDocumentSyncKind.Full) {
+        const document = await command.getTextDocument(this.session, id);
+        if (null != document) await this.session.merlin.sync(merlin.Sync.tell("start", "end", document.getText()), id);
+      }
+      const errors = await this.session.merlin.query(merlin.Query.errors(), id);
+      if (errors.class !== "return") return;
+      const diagnostics: types.Diagnostic[] = [];
```
Comment (jchavarri) [inline src/server/session/analyzer.ts]: This declaration triggers a `no-shadow-var` lint error in line 81 below, cause `diagnostic` is already declared here. There are a few other lint errors too, make sure you install TSLink extension :)

File: src/server/session/analyzer.ts [hunk 3]
```diff
@@ -46,6 +46,17 @@ export default class Analyzer implements rpc.Disposable {
...
+      for (const report of errors.value) diagnostics.push(await merlin.IErrorReport.intoCode(this.session, id, report));
+      this.session.connection.sendDiagnostics({ diagnostics, uri: id.uri });
```
Comment (jchavarri) [inline src/server/session/analyzer.ts]: I'm confused about this. How can running merlin diagnostic at all times fix the problem outlined in reasonml-editor/vscode-reasonml#45?

**Comment by jchavarri** [comment]
Thanks @wilonth !

> Let Merlin diagnostics go first, bsb's diagnostics may then override it when the file is saved

I'm afraid i'm not sure I understand what's the purpose. Issue https://github.com/reasonml-editor/vscode-reasonml/issues/45 was about stale Merlin errors after deleting a file. One of the benefits of setting `reason.bsb.enable` is that with `bsb` we can keep track of these errors and clear them when there's a close operation. So, in a way, that issue was somehow fixed with #32.

In this branch, when `reason.bsb.enable` is `false`, the errors are still stale after deleting a file:
stale-errors

> Catch External Preprocessor error (i.e from Refmt) in bsbOutput error parser

`bsb` is a different binary than `refmt`, so if we want to support parsing its output there should be a new file in `parser/` called `refmt` that spawn the process, parse output, etc. Could you elaborate on this part of the PR or provide an example of the output you want to parse, please?

**Comment by wilonth** [comment]
My bad I didn't test the new master code before making my modifications, what a late night hallucination. Indeed #32 fixed the issues already. I'll be more careful next time.

**Comment by freebroccolo** [comment]
No worries. Thanks for the attempt though.

**Pull request closed without merging**